src/canister_tests/src/tests.rs

```diff
@@ -75,7 +58,306 @@ fn registration_with_mismatched_sender_fails() {
+
+        assert!(matches!(
+            verification_response,
+            types::VerifyTentativeDeviceResponse::Verified
+        ));
+        Ok(())
+    }
+
+    /// Tests that the anchor info call returns information about tentative devices.
+    /// This enables the front-end to continue an in progress flow (e.g. after a refresh of the page).
+    #[test]
+    fn anchor_info_should_return_tentative_device() -> Result<(), CallError> {
+        let env = StateMachine::new();
+        let canister_id = framework::install_ii_canister(&env, framework::II_WASM_PREVIOUS.clone());
+        let user_number = flows::register_anchor(&env, canister_id);
+
+        api::enter_device_registration_mode(&env, canister_id, principal_1(), user_number)?;
+        let new_device = device_data_2();
+        api::add_tentative_device(
+            &env,
+            canister_id,
+            principal_2(),
+            user_number,
+            new_device.clone(),
+        )?;
+        let anchor_info = api::get_anchor_info(&env, canister_id, principal_1(), user_number)?;
+
+        assert!(matches!(
+            anchor_info.device_registration,
+            Some(types::DeviceRegistrationInfo {
+                tentative_device: Some(tenative_device),
+                ..
+            }) if tenative_device == new_device
+        ));
+        Ok(())
+    }
+
+    /// Tests that devices cannot be registered tentatively if the registration mode is not enabled.
+    #[test]
+    fn reject_tentative_device_if_not_in_registration_mode() -> Result<(), CallError> {
+        let env = StateMachine::new();
+        let canister_id = framework::install_ii_canister(&env, framework::II_WASM_PREVIOUS.clone());
+        let user_number = flows::register_anchor(&env, canister_id);
+
+        api::enter_device_registration_mode(&env, canister_id, principal_1(), user_number)?;
+        api::exit_device_registration_mode(&env, canister_id, principal_1(), user_number)?;
+        let result = api::add_tentative_device(
+            &env,
+            canister_id,
+            principal_2(),
+            user_number,
+            device_data_2(),
+        )?;
+
+        assert!(matches!(
+            result,
+            types::AddTentativeDeviceResponse::DeviceRegistrationModeOff
+        ));
+        Ok(())
+    }
+
+    /// Tests device registration mode expiration.
+    #[test]
+    fn reject_tentative_device_if_registration_mode_is_expired() -> Result<(), CallError> {
+        const REGISTRATION_MODE_EXPIRATION: Duration = Duration::from_secs(900);
+        let env = StateMachine::new();
+        let canister_id = framework::install_ii_canister(&env, framework::II_WASM_PREVIOUS.clone());
+        let user_number = flows::register_anchor(&env, canister_id);
+
+        api::enter_device_registration_mode(&env, canister_id, principal_1(), user_number)?;
+        env.set_time(
+            SystemTime::UNIX_EPOCH + REGISTRATION_MODE_EXPIRATION + Duration::from_secs(1),
+        );
```

Comment:
would be great if there was a way to modify the time instead of setting the time; here for instance we have no guarantee that we aren't going back in time (although unlikely). @roman-kashitsyn thoughts on that?